linchpin/provision/library/bkr_info.py

```diff
@@ -12,91 +11,82 @@
         self.conf.load_from_file(default_config)
         self.hub = HubProxy(logger=logger, conf=self.conf)
 
-    def _get_url(self, bkr_id):
-        """
-        Constructs the Beaker URL for the job related to the provided Beaker
-        ID. That ID should be all numeric, unless the structure of Beaker
-        changes in the future. If that's the case, then the ID should be
-        appropriately URL encoded to be appended to the end of a URL properly.
-        """
-        base = self.conf.get('HUB_URL', '')
-        if base == '':
-            raise Exception("Unable to construct URL")
-        if base[-1] != '/':
-            base += '/'
-        return base + 'jobs/' + bkr_id
-
     def get_system_statuses(self):
         """
         Checks on the status of a set of Beaker jobs (ids) and returns their
         hostname once the jobs have reached their defined status.
         """
         attempts = 0
-        pass_count = 0
-        all_count = len(self.ids)
         while attempts < self.max_attempts:
-            job_results = self._check_jobs(self.ids)
+            job_results, all_count = self._check_jobs()
             pass_count = 0
-            for resource in job_results['resources']:
+            for resource in job_results:
                 result = resource['result']
                 status = resource['status']
                 print >> stderr, "status: %s, result: %s" % (status, result)
                 if status not in ['Cancelled', 'Aborted']:
-                    if (result == 'Pass' or
-                       (result == 'Warn' and self.skip_no_system)):
+                    if result == 'Pass':
                         pass_count += 1
                     elif result in ['Fail', 'Warn', 'Panic', 'Completed']:
                         raise Exception("System failed with state"
                                         " '{0}'".format(result))
                 elif status == 'Aborted':
-                    if result == 'Warn' and self.skip_no_system:
-                        pass_count += 1
-                    else:
-                        raise Exception("System aborted")
+                    raise Exception("System aborted")
                 elif status == 'Cancelled':
                     raise Exception("System canceled")
             attempts += 1
             if pass_count == all_count:
-                return job_results['resources']
-            sleep(WAIT_TIME)
-        raise Exception("{0} system(s) never completed in {1}"
-                        " polling attempts. {2}".format(all_count - pass_count,
-                                                        attempts,
-                                                        dumps(job_results)))
+                return job_results
+            sleep(self.wait_time)
 
+        # max attempts exceeded, cancel jobs and fail
+        for job_id in _jprefix(self.ids):
+            self.hub.taskactions.stop(job_id, 'cancel',
```

Comment:
The one thing I would request as part of this changeset is that the cancellation of the beaker job also be attempted in the case of abort or cancel.

I think it would be best if the net result of a beaker provisioning transaction is either the job is running, the job was never submitted, or the job has been canceled.

I discussed this / the motivation for my reasoning in:
https://github.com/CentOS-PaaS-SIG/linchpin/issues/394#issuecomment-363466239